crates/derivation/src/batch_queue.rs

```diff
@@ -0,0 +1,47 @@
+use ethers_core::{types::Transaction, utils::rlp::decode};
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, VecDeque};
+
+use super::batch::Batch;
+use core::prelude::*;
+
+#[derive(Default, Debug)]
+pub struct BatchQueue {
+	l1_blocks: VecDeque<L1BlockRef>,
+	// Map batch timestamp to batches in order that they were received
+	batches: HashMap<u64, VecDeque<Batch>>,
+}
+
+const L2_BLOCK_TIME: u64 = 2u64;
+// const SEQ_WINDOW_SIZE: u64 = 3600u64;
+
+impl BatchQueue {
+	pub fn load_batches(&mut self, batches: impl Iterator<Item = Batch>, l1_origin: L1BlockRef) {
+		self.l1_blocks.push_back(l1_origin);
+		for b in batches {
+			println!("{b:?}");
+			if let Entry::Vacant(e) = self.batches.entry(b.batch.timestamp) {
+				e.insert(VecDeque::default());
+			}
+			self.batches.get_mut(&b.batch.timestamp).unwrap().push_back(b);
```

Comment:
Avoid the unwrap here by bubbling up an `eyre::Result`

Reply:
I need to use a different idiom here. The vec should be initialized there